src/privatesend-client.cpp

```diff
@@ -1268,22 +1269,23 @@ bool CPrivateSendClient::CreateDenominated(const CompactTallyItem& tallyItem, bo
 
             int nOutputs = 0;
 
-            // add each output up to 10 times until it can't be added again
+            // add each output up to 11 times until it can't be added again
             while(nValueLeft - nDenomValue >= 0 && nOutputs <= 10) {
                 CScript scriptDenom;
                 CPubKey vchPubKey;
                 //use a unique change address
-                assert(reservekeyDenom.GetReservedKey(vchPubKey)); // should never fail, as we just unlocked
+                std::shared_ptr<CReserveKey> reservekeyDenom = std::make_shared<CReserveKey>(pwalletMain);
+                reservekeyDenomVec.push_back(reservekeyDenom);
+
+                assert(reservekeyDenom->GetReservedKey(vchPubKey)); // should never fail, as we just unlocked
                 scriptDenom = GetScriptForDestination(vchPubKey.GetID());
-                // TODO: do not keep reservekeyDenom here
-                reservekeyDenom.KeepKey();
 
                 vecSend.push_back((CRecipient){ scriptDenom, nDenomValue, false });
 
                 //increment outputs and subtract denomination amount
                 nOutputs++;
                 nValueLeft -= nDenomValue;
-                LogPrintf("CreateDenominated1: nOutputsTotal: %d, nOutputs: %d, nValueLeft: %f\n", nOutputsTotal, nOutputs, (float)nValueLeft/COIN);
+                LogPrintf("CreateDenominated1: nOutputsTotal: %d, nOutputs: %d, nValueLeft: %f\n", nOutputsTotal + nOutputs, nOutputs, (float)nValueLeft/COIN);
```

Comment:
the idea here was to show the actual number of total outputs created.
now logs look like this:
```
2017-05-19 16:50:04 CreateDenominated1: nOutputsTotal: 0, nOutputs: 1, nValueLeft: 999.984985
2017-05-19 16:50:04 CreateDenominated1: nOutputsTotal: 0, nOutputs: 2, nValueLeft: 999.974976
2017-05-19 16:50:04 CreateDenominated1: nOutputsTotal: 0, nOutputs: 3, nValueLeft: 999.965027
2017-05-19 16:50:04 CreateDenominated1: nOutputsTotal: 0, nOutputs: 4, nValueLeft: 999.955017
2017-05-19 16:50:04 CreateDenominated1: nOutputsTotal: 0, nOutputs: 5, nValueLeft: 999.945007
2017-05-19 16:50:04 CreateDenominated1: nOutputsTotal: 0, nOutputs: 6, nValueLeft: 999.934998
2017-05-19 16:50:04 CreateDenominated1: nOutputsTotal: 0, nOutputs: 7, nValueLeft: 999.924988
2017-05-19 16:50:05 CreateDenominated1: nOutputsTotal: 0, nOutputs: 8, nValueLeft: 999.914978
2017-05-19 16:50:05 CreateDenominated1: nOutputsTotal: 0, nOutputs: 9, nValueLeft: 999.905029
2017-05-19 16:50:05 CreateDenominated1: nOutputsTotal: 0, nOutputs: 10, nValueLeft: 999.895020
2017-05-19 16:50:05 CreateDenominated1: nOutputsTotal: 0, nOutputs: 11, nValueLeft: 999.885010
2017-05-19 16:50:05 CreateDenominated1: nOutputsTotal: 11, nOutputs: 1, nValueLeft: 999.784973
```
you are right that now log is incorrect - variable nOutputsTotal has diffrent value than printed.
So maybe something like in may next commit?